Add support for adding/dropping indexes concurrently with Postgres

whatyouhide opened the pull request:
I added support for adding/dropping indexes concurrently as specified in #391.

I'm not sure this PR is ready to merge, I think I need a short review first.

The most straightforward part has been adding `@disable_ddl_transaction` to `Ecto.Migration`: what I did was adding a `@before_compile` hook in `Ecto.Migration.__using__/1` in order to define a `__disable_ddl_transaction__/0` function on the migration module. This way, I can retrieve that value in `Ecto.Migrator` and avoid wrapping everything inside a transaction if that's the case.

The part I'm the least sure about is actually adding the `CONCURRENTLY` keyword to the query. I added the `:concurrently` field to the `Ecto.Migration.Index` struct and added the `CONCURRENTLY` only for the Postgres adapter (since only Postgres supports this option AFAIK), but didn't take any measure to ensure that the `:concurrently` field is ignored (or raised against) by other adapters.

What do you think for now?

**PS** I'd also love a small review to the unit test I wrote; I have to check that no transactions are run, but the schema migrator runs one so I can't just do `refute_receive {:transaction, _}`. Let me know!

josevalim commented:
This pull request is **great**! I am adding just some tiny comments/suggestions!

josevalim commented on lib/ecto/migration.ex:
```diff
@@ -96,6 +107,15 @@ defmodule Ecto.Migration do
+  @doc false
+  defmacro __before_compile__(_env) do
+    quote do
+      def __disable_ddl_transaction__, do: @disable_ddl_transaction
```
What if we store this under the __migration__ function?

josevalim commented on lib/ecto/migrator.ex:
```diff
@@ -56,13 +56,18 @@ defmodule Ecto.Migrator do
       SchemaMigration.up(repo, version)
     end
+
+    if module.__disable_ddl_transaction__ do
```
We can move this to a function. `run_repo` or something.

josevalim commented on test/ecto/migrator_test.exs:
```diff
@@ -203,6 +212,18 @@ defmodule Ecto.MigratorTest do
+      # Assert there's only one transaction message, which is for when the
+      # SchemaMigration does his thing. If @disable_ddl_transaction was set to
+      # false, we would have *two* {:transaction, _} messages.
+      {:messages, messages} = Process.info(self, :messages)
```
You could do:

```elixir
# From the SchemaMigration insert transaction
assert_received {:transaction, _}

# No transaction from migration
refute_received {:transaction, _}
```

In fact, maybe we should add a test showing that repos with DDL transaction do run two transactions?

josevalim commented:
We just need one extra test here:

https://github.com/elixir-lang/ecto/blob/master/test/ecto/adapters/postgres_test.exs

That's where you can test if concurrently is being properly respected. :)

whatyouhide commented:
Ok, working on it!

whatyouhide commented:
Hey @josevalim, I made the changes you suggested. Quick wrap-up:

- Removed `__disable_ddl_transaction__` in favour of using `__migration__`
- Better tests for the transaction
- Added tests for the generated SQL
- Added `CONCURRENTLY` support for dropping indexes too (I completely missed that :|)

Let me know what else I can do :smiley:.

josevalim commented on lib/ecto/migration.ex:
```diff
@@ -69,8 +69,19 @@ defmodule Ecto.Migration do
+      name: atom,
+      columns: [atom | String.t],
+      unique: boolean,
+      concurrently: boolean
```
We need to update the docs of the `index` function too, right? Maybe add a couple paragraphs about how to use concurrent indexes and disable_ddl_transaction.

whatyouhide commented:
I'm usually a docs nerd, I completely forgot to add them here!

Last doubt: as the Postgres docs say, you can't create/drop indexes concurrently inside a transaction. Should we prevent setting the `:concurrently` option to `true` when `@disable_ddl_transaction` has not been set to `true`? We could just raise or we could let the error go down to Postgres/Postgrex. What do you think?

josevalim commented on test/ecto/migrator_test.exs:
```diff
@@ -45,6 +45,25 @@ defmodule Ecto.MigratorTest do
+    end
+  end
+
+  defmodule TransactionMigration do
```
Can we use an existing migration? Avoid defining many modules as they become hard to maintain and affect the testing time considerably. :)

whatyouhide replied:
Fixed this by `push --force`'ing so that the history looks niiice!

josevalim commented:
> Last doubt: as the Postgres docs say, you can't create/drop indexes concurrently inside a transaction. Should we prevent setting the :concurrently option to true when @disable_ddl_transaction has not been set to true? We could just raise or we could let the error go down to Postgres/Postgrex. What do you think?

I am not sure. Right now I am more inclined to not do anything, because maybe it is a postgrex only limitation or maybe it will even change with time. I think this path will be fine as long as the docs are clear. :)

whatyouhide commented:
Great, thanks! :smiley: :heart:
